staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -29,6 +29,44 @@ type Token interface {
 	AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
 }
 
+func WrapInInstrumentedRequest(req Request, authenticatorName string) *InstrumentedRequest {
+	return &InstrumentedRequest{Name: authenticatorName, delegate: req}
+}
+
+// InstrumentedRequest wraps Request and allows instrumentation to identify authenticators via the Name field.
+type InstrumentedRequest struct {
+	// Name identifies the delegate authenticator.
+	Name     string
+	delegate Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	if err != nil {
+		err = InstrumentedError{Name: i.Name, delegate: err}
+	}
+
+	if resp != nil {
```

Comment:
Should this only set the AuthenticatorName if the request is authenticated?

Reply:
Good question: I would imagine we want to instrument requests that failed authentication as well. So, not checking for "ok" seems reasonable. WDYT?

Reply:
If it failed authentication, that really means it failed authentication by all authenticators, right? Or would this be for the case where it provided a specific type of credential, but it was invalid?

Reply:
To elaborate, if I make a request without any credentials, I don't want the last (or first) authenticator in the chain to be attributed the failure, otherwise that will be misleading.

Reply:
As written, the failure (in the scenario where all authenticators in a union fail) will be attributed to the "unknown" authenticator, which is not what we want (see ""union intersection - all authenticators fail" test-case). 
This is due to [AuthenticateRequest](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go#L56) being delegated to the "union" authenticator which is not instrumented.

One option is to instrument the "union" authenticators (request and token).

An alternative would be to instrument InstrumentedAuthenticator, InstrumentedToken and InstrumentedPassword, which seems logical. 
WDYT?

Reply:
Ended-up instrumenting InstrumentedAuthenticator, InstrumentedToken and InstrumentedPassword - I believe we now get the expected behaviour.